mm2src/crypto/src/crypto_ctx.rs

```diff
@@ -41,9 +40,32 @@ impl From<PrivKeyError> for CryptoInitError {
     fn from(e: PrivKeyError) -> Self { CryptoInitError::InvalidPassphrase(e) }
 }
 
-pub enum CryptoCtx {
-    KeyPair(KeyPairArc),
-    HardwareWallet(HardwareWalletArc),
+#[derive(Debug)]
+pub enum HwCtxInitError<ProcessorError> {
+    InitializedAlready,
+    HwError(HwError),
+    ProcessorError(ProcessorError),
+}
+
+impl<ProcessorError> From<HwProcessingError<ProcessorError>> for HwCtxInitError<ProcessorError> {
+    fn from(e: HwProcessingError<ProcessorError>) -> Self {
+        match e {
+            HwProcessingError::HwError(hw_error) => HwCtxInitError::HwError(hw_error),
+            HwProcessingError::ProcessorError(processor_error) => HwCtxInitError::ProcessorError(processor_error),
+        }
+    }
+}
+
+/// This is required for converting `MmError<HwProcessingError<E>>` into `MmError<InitHwCtxError<E>>`.
+impl<E> NotSame for HwCtxInitError<E> {}
+
+pub struct CryptoCtx {
+    iguana_ctx: IguanaArc,
+    /// Can be initialized on [`CryptoCtx::init_hw_ctx_with_trezor`].
+    /// Please note that it's preferred to use `AsyncRwLock` here
+    /// because [`CryptoCtx::hw_ctx`] can be locked for a long time while `HardwareWalletCtx` is initializing,
+    /// and `AsyncRwLock` allows a runtime to poll other futures even on the same thread.
+    hw_ctx: AsyncRwLock<Option<HardwareWalletArc>>,
```

Comment:
We should be able to initialize `CryptoCtx::hw_ctx` on `init_trezor` RPC and probably reset the field to `None` if the user desire. But to avoid double initialization of Hardware Wallet Context we need to hold a mutex locked.

Reply:
Can we have an additional state like `enum CryptoCtxInitState { NotInitialized, InProgress, Initialized }` instead? Double initialization attempt will fail if `matches!(current_state, InProgress | Initialized)`.

Reply:
Ooh, I see, that makes sense! Going to do this right now

Reply:
Done, please check it :)